File: internal/validation/validation_test.go
package validation

import (
	"math"
	"testing"
)

// Test cases in this file have been validated against the actual Python is_number() function
// from the source code (Simple Calculator by Python.py, lines 19-34) to ensure exact behavioral parity.
// Key findings from Python validation:
//   - Simple integers: "123", "-123", "+123" → valid
//   - Unsigned floats: "123.456", "0.123" → valid
//   - Signed floats with leading zero: "-0.123", "+0.123" → valid
//   - Leading decimals: ".123", "-.123", "+.123" → valid
//   - Space-prefixed: " 123", " .123", " 0.123" → valid
//   - Signed floats WITHOUT leading zero: "-123.456", "+123.456" → INVALID (Python limitation)
// These test cases mirror the Python behavior exactly.

// TestIsValidNumber_ValidIntegers tests validation of valid integer formats
func TestIsValidNumber_ValidIntegers(t *testing.T) {
	testCases := []string{
		"123",
		"-123",
		"+123",
		"0",
		"1",
		"999999",
	}

	for _, tc := range testCases {
		t.Run(tc, func(t *testing.T) {
			if !IsValidNumber(tc) {
				t.Errorf("IsValidNumber(%q) = false, want true", tc)
			}
		})
	}
}

// TestIsValidNumber_ValidFloats tests validation of valid float formats
func TestIsValidNumber_ValidFloats(t *testing.T) {
	testCases := []string{
		"123.456",
		"0.123",
		"-0.123",
		"+0.123",
		"123.0",
		"0.0",
	}

	for _, tc := range testCases {
		t.Run(tc, func(t *testing.T) {
			if !IsValidNumber(tc) {
				t.Errorf("IsValidNumber(%q) = false, want true", tc)
			}
		})
	}
}

// TestIsValidNumber_LeadingDecimal tests validation of leading decimal notation
func TestIsValidNumber_LeadingDecimal(t *testing.T) {
	testCases := []string{
		".123",
		"-.123",
		"+.123",
		".0",
		".999",
	}

	for _, tc := range testCases {
		t.Run(tc, func(t *testing.T) {
			if !IsValidNumber(tc) {
				t.Errorf("IsValidNumber(%q) = false, want true", tc)
			}
		})
	}
}

// TestIsValidNumber_SpacePrefixed tests validation of space-prefixed numeric strings
// The Python source code checks s[0] in ['-', '+', '.', '0', ' '], allowing space-prefixed numbers
func TestIsValidNumber_SpacePrefixed(t *testing.T) {
	testCases := []string{
		" 123",   // space then integer
		" .123",  // space then leading decimal
		" 0.123", // space then zero decimal
	}

	for _, tc := range testCases {
		t.Run(tc, func(t *testing.T) {
			if !IsValidNumber(tc) {
				t.Errorf("IsValidNumber(%q) = false, want true", tc)
			}
		})
	}
}

// TestIsValidNumber_Invalid tests rejection of invalid formats
func TestIsValidNumber_Invalid(t *testing.T) {
	testCases := []string{
		"",         // empty string
		" ",        // single space
		"abc",      // non-numeric
		"12a34",    // mixed
		"12.34.56", // multiple decimals
		"--123",    // double negative
		"++123",    // double positive
		"+-123",    // mixed signs
		".",        // just decimal point
		"-",        // just minus sign
		"+",        // just plus sign
		"12 34",    // space in middle
		"12.34a",   // letter at end
		"a12.34",   // letter at start
		"-123.456", // signed float (not supported by Python is_number)
		"+123.456", // signed float (not supported by Python is_number)
	}

	for _, tc := range testCases {
		t.Run(tc, func(t *testing.T) {
			if IsValidNumber(tc) {
				t.Errorf("IsValidNumber(%q) = true, want false", tc)
			}
		})
	}
}

// TestParseNumber_Integers tests parsing of integer formats
func TestParseNumber_Integers(t *testing.T) {
	testCases := []struct {
		input       string
		expectedVal float64
		expectedInt bool
	}{
		{"123", 123.0, true},
		{"-123", -123.0, true},
		{"+123", 123.0, true},
		{"0", 0.0, true},
	}

	for _, tc := range testCases {
		t.Run(tc.input, func(t *testing.T) {
			val, isInt, err := ParseNumber(tc.input)
			if err != nil {
				t.Fatalf("ParseNumber(%q) returned error: %v", tc.input, err)
			}
			if val != tc.expectedVal {
				t.Errorf("ParseNumber(%q) value = %v, want %v", tc.input, val, tc.expectedVal)
			}
			if isInt != tc.expectedInt {
				t.Errorf("ParseNumber(%q) isInt = %v, want %v", tc.input, isInt, tc.expectedInt)
			}
		})
	}
}

// TestParseNumber_Floats tests parsing of float formats
func TestParseNumber_Floats(t *testing.T) {
	testCases := []struct {
		input       string
		expectedVal float64
		expectedInt bool
	}{
		{"123.456", 123.456, false},
		{"0.123", 0.123, false},
		{"-0.123", -0.123, false},
		{"+0.123", 0.123, false},
		{".123", 0.123, false},
		{"-.123", -0.123, false},
		{"+.123", 0.123, false},
		{"123.0", 123.0, false},
	}

	for _, tc := range testCases {
		t.Run(tc.input, func(t *testing.T) {
			val, isInt, err := ParseNumber(tc.input)
			if err != nil {
				t.Fatalf("ParseNumber(%q) returned error: %v", tc.input, err)
			}
			if math.Abs(val-tc.expectedVal) > 0.0001 {
				t.Errorf("ParseNumber(%q) value = %v, want %v", tc.input, val, tc.expectedVal)
			}
			if isInt != tc.expectedInt {
				t.Errorf("ParseNumber(%q) isInt = %v, want %v", tc.input, isInt, tc.expectedInt)
			}
		})
	}
}

// TestParseNumber_InvalidInput tests error handling for invalid inputs
func TestParseNumber_InvalidInput(t *testing.T) {
	testCases := []string{
		"abc",
		"",
		"12.34.56",
		"not a number",
	}

	for _, tc := range testCases {
		t.Run(tc, func(t *testing.T) {
			_, _, err := ParseNumber(tc)
			if err == nil {
				t.Errorf("ParseNumber(%q) expected error, got nil", tc)
			}
		})
	}
}

// TestParseNumber_PreservesIntFloatDistinction verifies that the function
// preserves the distinction between integer and float notation
func TestParseNumber_PreservesIntFloatDistinction(t *testing.T) {
	// "123" should be marked as integer
	val1, isInt1, _ := ParseNumber("123")
	if !isInt1 {
		t.Errorf("ParseNumber(\"123\") should indicate integer format")
	}
	if val1 != 123.0 {
		t.Errorf("ParseNumber(\"123\") value = %v, want 123.0", val1)
	}

	// "123.0" should be marked as float
	val2, isInt2, _ := ParseNumber("123.0")
	if isInt2 {
		t.Errorf("ParseNumber(\"123.0\") should indicate float format")
	}
	if val2 != 123.0 {
		t.Errorf("ParseNumber(\"123.0\") value = %v, want 123.0", val2)
	}
}

// TestValidationWorkflow tests the complete validation and parsing workflow
func TestValidationWorkflow(t *testing.T) {
	testCases := []struct {
		input       string
		shouldValid bool
		expectedVal float64
		expectedInt bool
	}{
		{"123", true, 123.0, true},
		{"123.456", true, 123.456, false},
		{".123", true, 0.123, false},
		{"-.123", true, -0.123, false},
		{"", false, 0, false},
		{"abc", false, 0, false},
	}

	for _, tc := range testCases {
		t.Run(tc.input, func(t *testing.T) {
			// First validate
			isValid := IsValidNumber(tc.input)
			if isValid != tc.shouldValid {
				t.Errorf("IsValidNumber(%q) = %v, want %v", tc.input, isValid, tc.shouldValid)
			}

			// If should be valid, parse it
			if tc.shouldValid {
				val, isInt, err := ParseNumber(tc.input)
				if err != nil {
					t.Fatalf("ParseNumber(%q) returned error: %v", tc.input, err)
				}
				if math.Abs(val-tc.expectedVal) > 0.0001 {
					t.Errorf("ParseNumber(%q) value = %v, want %v", tc.input, val, tc.expectedVal)
				}
				if isInt != tc.expectedInt {
					t.Errorf("ParseNumber(%q) isInt = %v, want %v", tc.input, isInt, tc.expectedInt)
				}
			}
		})
	}
}

// TestValidateNumber tests the error-returning validation function
func TestValidateNumber(t *testing.T) {
	// Test valid inputs
	validInputs := []string{"123", "123.456", ".123", "-.123", "-0.123"}
	for _, input := range validInputs {
		t.Run("valid_"+input, func(t *testing.T) {
			err := ValidateNumber(input)
			if err != nil {
				t.Errorf("ValidateNumber(%q) returned error: %v, want nil", input, err)
			}
		})
	}

	// Test invalid inputs
	invalidInputs := []string{"", " ", "abc", "12.34.56"}
	for _, input := range invalidInputs {
		t.Run("invalid_"+input, func(t *testing.T) {
			err := ValidateNumber(input)
			if err == nil {
				t.Errorf("ValidateNumber(%q) returned nil, want error", input)
			}
			if err != nil && err.Error() != ValidationError {
				t.Errorf("ValidateNumber(%q) error = %q, want %q", input, err.Error(), ValidationError)
			}
		})
	}
}

// TestValidateAndParseNumber tests the convenience function
func TestValidateAndParseNumber(t *testing.T) {
	testCases := []struct {
		input       string
		shouldValid bool
		expectedVal float64
		expectedInt bool
	}{
		{"123", true, 123.0, true},
		{"123.456", true, 123.456, false},
		{".123", true, 0.123, false},
		{"-.123", true, -0.123, false},
		{"", false, 0, false},
		{"abc", false, 0, false},
	}

	for _, tc := range testCases {
		t.Run(tc.input, func(t *testing.T) {
			val, isInt, err := ValidateAndParseNumber(tc.input)

			if tc.shouldValid {
				if err != nil {
					t.Fatalf("ValidateAndParseNumber(%q) returned error: %v", tc.input, err)
				}
				if math.Abs(val-tc.expectedVal) > 0.0001 {
					t.Errorf("ValidateAndParseNumber(%q) value = %v, want %v", tc.input, val, tc.expectedVal)
				}
				if isInt != tc.expectedInt {
					t.Errorf("ValidateAndParseNumber(%q) isInt = %v, want %v", tc.input, isInt, tc.expectedInt)
				}
			} else {
				if err == nil {
					t.Errorf("ValidateAndParseNumber(%q) returned nil error, want validation error", tc.input)
				}
				if err != nil && err.Error() != ValidationError {
					t.Errorf("ValidateAndParseNumber(%q) error = %q, want %q", tc.input, err.Error(), ValidationError)
				}
			}
		})
	}
}

File: src/core/validation.test.ts
import { describe, it, expect } from 'vitest';
import { isNumber, castToNumber } from './validation';

describe('isNumber', () => {
  describe('valid integer formats', () => {
    it('should accept positive integers', () => {
      expect(isNumber('123')).toBe(true);
      expect(isNumber('1')).toBe(true);
      expect(isNumber('999999')).toBe(true);
    });

    it('should accept negative integers', () => {
      expect(isNumber('-123')).toBe(true);
      expect(isNumber('-1')).toBe(true);
      expect(isNumber('-999')).toBe(true);
    });

    it('should accept positive integers with plus sign', () => {
      expect(isNumber('+123')).toBe(true);
      expect(isNumber('+1')).toBe(true);
      expect(isNumber('+999')).toBe(true);
    });

    it('should accept zero', () => {
      expect(isNumber('0')).toBe(true);
      expect(isNumber('+0')).toBe(true);
      expect(isNumber('-0')).toBe(true);
    });

    it('should accept integers with leading space', () => {
      expect(isNumber(' 123')).toBe(true);
      expect(isNumber(' 1')).toBe(true);
    });
  });

  describe('valid float formats', () => {
    it('should accept floats with leading zero', () => {
      expect(isNumber('0.123')).toBe(true);
      expect(isNumber('0.1')).toBe(true);
      expect(isNumber('0.999')).toBe(true);
    });

    it('should accept floats without leading zero', () => {
      expect(isNumber('.123')).toBe(true);
      expect(isNumber('.1')).toBe(true);
      expect(isNumber('.999')).toBe(true);
    });

    it('should accept signed floats with leading zero', () => {
      expect(isNumber('-0.123')).toBe(true);
      expect(isNumber('+0.123')).toBe(true);
      expect(isNumber('-0.1')).toBe(true);
      expect(isNumber('+0.999')).toBe(true);
    });

    it('should accept signed floats without leading zero', () => {
      expect(isNumber('-.123')).toBe(true);
      expect(isNumber('+.123')).toBe(true);
      expect(isNumber('-.1')).toBe(true);
      expect(isNumber('+.999')).toBe(true);
    });

    it('should accept standard float formats', () => {
      expect(isNumber('123.456')).toBe(true);
      expect(isNumber('1.5')).toBe(true);
      expect(isNumber('999.999')).toBe(true);
    });

    it('should reject signed standard floats (Python limitation)', () => {
      // Python's is_number does NOT accept signed floats like "-123.456"
      // It only accepts unsigned floats "123.456" and special cases like "-0.123" or "-.123"
      expect(isNumber('-123.456')).toBe(false);
      expect(isNumber('+123.456')).toBe(false);
      expect(isNumber('-1.5')).toBe(false);
    });

    it('should accept floats with leading space and decimal', () => {
      expect(isNumber(' 0.123')).toBe(true);
    });
  });

  describe('invalid formats', () => {
    it('should reject empty string', () => {
      expect(isNumber('')).toBe(false);
    });

    it('should reject alphabetic characters', () => {
      expect(isNumber('abc')).toBe(false);
      expect(isNumber('1abc')).toBe(false);
      expect(isNumber('abc123')).toBe(false);
      expect(isNumber('12a34')).toBe(false);
    });

    it('should reject special characters', () => {
      expect(isNumber('!@#')).toBe(false);
      expect(isNumber('123!')).toBe(false);
      expect(isNumber('12@34')).toBe(false);
    });

    it('should reject multiple signs', () => {
      expect(isNumber('--123')).toBe(false);
      expect(isNumber('++123')).toBe(false);
      expect(isNumber('+-123')).toBe(false);
    });

    it('should reject multiple decimal points', () => {
      expect(isNumber('1.2.3')).toBe(false);
      // Note: Python implementation accepts "..123" due to its logic
      // expect(isNumber('..123')).toBe(false);
      expect(isNumber('1..23')).toBe(false);
    });

    it('should reject decimal point only', () => {
      expect(isNumber('.')).toBe(false);
      expect(isNumber('-.')).toBe(false);
      expect(isNumber('+.')).toBe(false);
    });

    it('should reject sign only', () => {
      expect(isNumber('-')).toBe(false);
      expect(isNumber('+')).toBe(false);
    });

    it('should reject space only', () => {
      expect(isNumber(' ')).toBe(false);
    });

    it('should reject numbers with internal spaces', () => {
      expect(isNumber('12 34')).toBe(false);
      expect(isNumber('1 .23')).toBe(false);
    });
  });

  describe('edge cases', () => {
    it('should handle numbers starting with zero', () => {
      expect(isNumber('0123')).toBe(true);
      expect(isNumber('00.123')).toBe(true);
    });

    it('should handle numbers starting with space followed by decimal', () => {
      expect(isNumber(' .123')).toBe(true);
    });

    it('should reject incomplete float formats', () => {
      // Note: Python implementation accepts "123." because "123".isdigit() is True
      // expect(isNumber('123.')).toBe(false);
      expect(isNumber('-123.')).toBe(false);
      expect(isNumber('.0.')).toBe(false);
    });

    it('should handle single digit with decimal', () => {
      expect(isNumber('1.0')).toBe(true);
      expect(isNumber('0.0')).toBe(true);
    });

    it('should accept trailing decimal for integers (Python quirk)', () => {
      // Python's is_number accepts "123." because "123".isdigit() is True
      expect(isNumber('123.')).toBe(true);
      expect(isNumber('0.')).toBe(true);
    });

    it('should accept leading double decimal (Python quirk)', () => {
      // Python's is_number accepts "..123" because s[0]='.' and s[1]='.' and s[2:]='123' isdigit
      expect(isNumber('..123')).toBe(true);
    });
  });
});

describe('castToNumber', () => {
  describe('integer conversion', () => {
    it('should convert positive integers', () => {
      expect(castToNumber('123')).toBe(123);
      expect(castToNumber('1')).toBe(1);
      expect(castToNumber('999')).toBe(999);
    });

    it('should convert negative integers', () => {
      expect(castToNumber('-123')).toBe(-123);
      expect(castToNumber('-1')).toBe(-1);
      expect(castToNumber('-999')).toBe(-999);
    });

    it('should convert integers with plus sign', () => {
      expect(castToNumber('+123')).toBe(123);
      expect(castToNumber('+1')).toBe(1);
    });

    it('should convert zero', () => {
      expect(castToNumber('0')).toBe(0);
    });

    it('should convert integers with leading space', () => {
      expect(castToNumber(' 123')).toBe(123);
    });
  });

  describe('float conversion', () => {
    it('should convert floats with leading zero', () => {
      expect(castToNumber('0.123')).toBe(0.123);
      expect(castToNumber('0.5')).toBe(0.5);
    });

    it('should convert floats without leading zero', () => {
      expect(castToNumber('.123')).toBe(0.123);
      expect(castToNumber('.5')).toBe(0.5);
    });

    it('should convert signed floats', () => {
      expect(castToNumber('-0.123')).toBe(-0.123);
      expect(castToNumber('+0.123')).toBe(0.123);
      expect(castToNumber('-.123')).toBe(-0.123);
      expect(castToNumber('+.123')).toBe(0.123);
    });

    it('should convert standard float formats', () => {
      expect(castToNumber('123.456')).toBe(123.456);
      expect(castToNumber('-123.456')).toBe(-123.456);
      expect(castToNumber('1.5')).toBe(1.5);
    });

    it('should handle very small decimals', () => {
      expect(castToNumber('0.001')).toBe(0.001);
      expect(castToNumber('.001')).toBe(0.001);
    });
  });

  describe('type semantics', () => {
    it('should use parseFloat for numbers with decimal point', () => {
      const result = castToNumber('123.0');
      expect(result).toBe(123.0);
      expect(typeof result).toBe('number');
    });

    it('should use parseInt for numbers without decimal point', () => {
      const result = castToNumber('123');
      expect(result).toBe(123);
      expect(typeof result).toBe('number');
    });
  });

  describe('edge cases', () => {
    it('should handle numbers with leading zeros', () => {
      expect(castToNumber('0123')).toBe(123);
      expect(castToNumber('00.123')).toBe(0.123);
    });

    it('should handle negative zero', () => {
      expect(castToNumber('-0')).toBe(-0);
      expect(castToNumber('-0.0')).toBe(-0);
    });

    it('should handle space-prefixed numbers', () => {
      expect(castToNumber(' 123')).toBe(123);
      expect(castToNumber(' .123')).toBe(0.123);
      expect(castToNumber(' 0.123')).toBe(0.123);
    });
  });
});
